canvas/canvas.txt

```diff
@@ -0,0 +1,384 @@
+<repeatcount> = 17
+<repeatcount> = 65
+<repeatcount> = 257
+<repeatcount> = 1025
+<repeatcount> = 4097
+<repeatcount> = 65537
+<repeatstr> = String.fromCharCode(<int min=0 max=127>)
+<repeatstr> = String.fromCharCode(<int min=0 max=127>, <int min=0 max=127>)
+<repeatstr> = String.fromCharCode(<int min=0 max=127>, <int min=0 max=127>, <int min=0 max=127>)
+
+<color> = #<hex><hex><hex><hex><hex><hex>
+<color> = rgb(<int min=0 max=255>,<int min=0 max=255>,<int min=0 max=255>)
+
+<fillRule> = "nonzero"
+<fillRule> = "evenodd"
+
+<lineCap> = "butt"
+<lineCap> = "round"
+<lineCap> = "square"
+<lineJoin> = "bevel"
+<lineJoin> = "round"
+<lineJoin> = "miter"
+
+<textAlign> = "left"
+<textAlign> = "right"
+<textAlign> = "center"
+<textAlign> = "start"
+<textAlign> = "end"
+
+<textBaseline> = "top" 
+<textBaseline> = "hanging"
+<textBaseline> = "middle"
+<textBaseline> = "alphabetic"
+<textBaseline> = "ideographic"
+<textBaseline> = "bottom"
+
+<txtDirection> = "ltr"
+<txtDirection> = "rtl"
+<txtDirection> = "inherit"
+
+<repetition> = "repeat" 
+<repetition> = "repeat-x" 
+<repetition> = "repeat-y" 
+<repetition> = "no-repeat"
+
+<imageFormat> = "image/png"
+<imageFormat> = "image/webp"
+
+<mathOperation> = +
+<mathOperation> = -
+<mathOperation> = /
+<mathOperation> = *
+
+<globalCompositeOperation> = "source-over"
+<globalCompositeOperation> = "source-in"
+<globalCompositeOperation> = "source-out"
+<globalCompositeOperation> = "source-atop"
+<globalCompositeOperation> = "destination-over"
+<globalCompositeOperation> = "destination-in"
+<globalCompositeOperation> = "destination-out"
+<globalCompositeOperation> = "destination-atop"
+<globalCompositeOperation> = "lighter"
+<globalCompositeOperation> = "copy"
+<globalCompositeOperation> = "xor"
+<globalCompositeOperation> = "multiply"
+<globalCompositeOperation> = "screen"
+<globalCompositeOperation> = "overlay"
+<globalCompositeOperation> = "darken"
+<globalCompositeOperation> = "lighten"
+<globalCompositeOperation> = "color-dodge"
+<globalCompositeOperation> = "hard-light"
+<globalCompositeOperation> = "soft-light"
+<globalCompositeOperation> = "difference"
+<globalCompositeOperation> = "exclusion"
+<globalCompositeOperation> = "hue"
+<globalCompositeOperation> = "saturation"
+<globalCompositeOperation> = "color"
+<globalCompositeOperation> = "luminosity"
+
+<this> = this
+
+<root root=true> = <lines count=50>
+
+#!lineguard try { <line>; } catch(e) { console.log(e.message) }
+!varformat fuzzvar%05d
+!begin lines
+!include common.txt
+
+# Image
+<new img> = new Image();
+<new img> = new Image(<fuzzint>, <fuzzint>);
+<new img> = new Image(<largeint>, <largeint>);
+<new img> = new Image(<interestingint>, <interestingint>);
+<new img> = new Image(<this>, <interestingint>);
+<new img> = new Image(<interestingint>, <this>);
+<img>.src = "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///ywAAAAAAQABAAACAUwAOw==";
+
+# Uint8ClampedArray
+<new uintc8> = new Uint8ClampedArray(<fuzzint>);

```

Comment:
Instead of repeating the lines like this, you can declare a single integer type, e.g.

```
<canvasint> = <interestingint>
<canvasint> = <fuzzint>
<canvasint> = <largeint>
```

and then just write

`<new uintc8> = new Uint8ClampedArray(<canvasint>);`

Which avoids a lot of line duplication here and in other places.

Reply:
You are *so* right, I added so much redundant code and I ended up getting bored (programming is tough!) - I've cleaned up this mess!